src/sass_types.h

```diff
@@ -0,0 +1,225 @@
+
+      CoreValue(Sass_Value* v) : Value(v) {}
+
+      Sass_Value* get_sass_value() { 
+        return sass_clone_value(this->value); 
+      }
+
+      Local<Object> get_js_object() {
+        Local<Object> wrapper = NanNew(T::get_constructor())->NewInstance();
+
+        delete static_cast<T*>(NanGetInternalFieldPointer(wrapper, 0));
+        NanSetInternalFieldPointer(wrapper, 0, this);
+
+        return wrapper;
+      }
+
+      static Handle<Function> get_constructor() {
+        if (constructor.IsEmpty()) {
+          Local<FunctionTemplate> tpl = FunctionTemplate::New(New);
+          tpl->SetClassName(NanNew(T::get_constructor_name()));
+          tpl->InstanceTemplate()->SetInternalFieldCount(1);
+          T::initPrototype(tpl->PrototypeTemplate());
+          constructor = Persistent<Function>::New(tpl->GetFunction());
+        }
+
+        return constructor;
+      }
+
+      static NAN_METHOD(New) {
+        Sass_Value* value = T::construct(args);
+        T* obj = new T(value);
+        sass_delete_value(value);
+
+        NanSetInternalFieldPointer(args.This(), 0, obj);
+
+        return args.This();
+      }
+
+    protected:
+      static T* unwrap(Local<Object> obj) {
+        return static_cast<T*>(Factory::unwrap(obj));
+      }
+
+    private:
+      static Persistent<Function> constructor;
+  };
+
+
+  template <class T>
+  Persistent<Function> CoreValue<T>::constructor;
+
+
+  class Number : public CoreValue<Number> {
+    public:
+      Number(Sass_Value* v) : CoreValue(v) {}
+      static char const* get_constructor_name() { return "SassNumber"; }
+      static Sass_Value* construct(const Arguments&);
+
+      static void initPrototype(Handle<ObjectTemplate> proto) {
+        proto->Set(NanNew("getValue"), FunctionTemplate::New(GetValue)->GetFunction());
+        proto->Set(NanNew("getUnit"), FunctionTemplate::New(GetUnit)->GetFunction());
+        proto->Set(NanNew("setValue"), FunctionTemplate::New(SetValue)->GetFunction());
+        proto->Set(NanNew("setUnit"), FunctionTemplate::New(SetUnit)->GetFunction());
+      }
+
+      static NAN_METHOD(GetValue);
+      static NAN_METHOD(GetUnit);
+      static NAN_METHOD(SetValue);
+      static NAN_METHOD(SetUnit);
+  };
+
+
+  class String : public CoreValue<String> {
+    public:
+      String(Sass_Value* v) : CoreValue(v) {}
+      static char const* get_constructor_name() { return "SassString"; }
+      static Sass_Value* construct(const Arguments&);
+
+      static void initPrototype(Handle<ObjectTemplate> proto) {
+        proto->Set(NanNew("getValue"), FunctionTemplate::New(GetValue)->GetFunction());
+        proto->Set(NanNew("setValue"), FunctionTemplate::New(SetValue)->GetFunction());
+      }
+
+      static NAN_METHOD(GetValue);
+      static NAN_METHOD(SetValue);
+  };
+
+
+  class Color : public CoreValue<Color> {
+    public:
+      Color(Sass_Value* v) : CoreValue(v) {}
+      static char const* get_constructor_name() { return "SassColor"; }
```

Comment:
Since we are using C++11, you may want to declare it as:

```c++
constexpr char* constructor_name = "SassBoolean";
```

How about modifying *the must implemented by derived class* members with `virtual` inheritance?

Also, you may want to take a look at http://stackoverflow.com/a/16296215/863980.